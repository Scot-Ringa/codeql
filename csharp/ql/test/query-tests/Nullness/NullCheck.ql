import csharp
import semmle.code.csharp.controlflow.Guards

from DereferenceableExpr de, Guards::Guard reason
where de.guardSuggestsMaybeNull(reason)
select reason, de
