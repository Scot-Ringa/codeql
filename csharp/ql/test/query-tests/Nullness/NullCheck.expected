| Assert.cs:10:22:10:30 | ... != ... | Assert.cs:10:22:10:22 | access to local variable s |
| Assert.cs:14:23:14:23 | access to local variable s | Assert.cs:14:23:14:23 | access to local variable s |
| Assert.cs:18:26:18:26 | access to local variable s | Assert.cs:18:26:18:26 | access to local variable s |
| Assert.cs:22:23:22:31 | ... == ... | Assert.cs:22:23:22:23 | access to local variable s |
| Assert.cs:26:23:26:31 | ... != ... | Assert.cs:26:23:26:23 | access to local variable s |
| Assert.cs:30:24:30:32 | ... != ... | Assert.cs:30:24:30:24 | access to local variable s |
| Assert.cs:34:24:34:32 | ... == ... | Assert.cs:34:24:34:24 | access to local variable s |
| Assert.cs:38:23:38:31 | ... != ... | Assert.cs:38:23:38:23 | access to local variable s |
| Assert.cs:42:24:42:32 | ... == ... | Assert.cs:42:24:42:24 | access to local variable s |
| Assert.cs:46:23:46:31 | ... == ... | Assert.cs:46:23:46:23 | access to local variable s |
| Assert.cs:50:24:50:32 | ... != ... | Assert.cs:50:24:50:24 | access to local variable s |
| B.cs:12:13:12:32 | call to operator == | B.cs:12:13:12:24 | access to local variable eqCallAlways |
| B.cs:12:13:12:32 | call to operator == | B.cs:12:29:12:32 | null |
| B.cs:15:13:15:22 | call to operator != | B.cs:15:13:15:14 | access to local variable b2 |
| B.cs:15:13:15:22 | call to operator != | B.cs:15:19:15:22 | null |
| B.cs:18:13:18:22 | call to operator == | B.cs:18:13:18:14 | access to local variable b3 |
| B.cs:18:13:18:22 | call to operator == | B.cs:18:19:18:22 | null |
| B.cs:22:13:22:33 | call to operator != | B.cs:22:13:22:25 | access to local variable neqCallAlways |
| B.cs:22:13:22:33 | call to operator != | B.cs:22:30:22:33 | null |
| B.cs:53:17:53:33 | ... != ... | B.cs:53:17:53:25 | (...) ... |
| B.cs:53:17:53:33 | ... != ... | B.cs:53:30:53:33 | null |
| B.cs:55:26:55:36 | call to method Equals | B.cs:55:26:55:26 | access to local variable o |
| B.cs:55:26:55:36 | call to method Equals | B.cs:55:35:55:35 | access to local variable o |
| C.cs:11:19:11:27 | ... == ... | C.cs:11:19:11:19 | access to local variable o |
| C.cs:11:19:11:27 | ... == ... | C.cs:11:24:11:27 | null |
| C.cs:16:15:16:23 | ... != ... | C.cs:16:15:16:15 | access to local variable o |
| C.cs:16:15:16:23 | ... != ... | C.cs:16:20:16:23 | null |
| C.cs:24:13:24:21 | ... != ... | C.cs:24:13:24:13 | access to parameter o |
| C.cs:28:37:28:45 | ... == ... | C.cs:28:37:28:37 | access to parameter o |
| C.cs:30:40:30:48 | ... != ... | C.cs:30:40:30:40 | access to parameter o |
| C.cs:34:13:34:21 | ... == ... | C.cs:34:13:34:13 | access to parameter o |
| C.cs:41:22:41:30 | ... == ... | C.cs:41:22:41:22 | access to local variable s |
| C.cs:45:22:45:30 | ... != ... | C.cs:45:22:45:22 | access to local variable s |
| C.cs:56:23:56:24 | access to local variable o2 | C.cs:56:23:56:24 | access to local variable o2 |
| C.cs:71:26:71:27 | access to local variable o3 | C.cs:71:26:71:27 | access to local variable o3 |
| C.cs:113:22:113:36 | ... == ... | C.cs:113:22:113:28 | access to local variable colours |
| C.cs:113:22:113:36 | ... == ... | C.cs:113:33:113:36 | null |
| C.cs:120:13:120:28 | ... == ... | C.cs:120:13:120:20 | access to local variable children |
| C.cs:120:13:120:28 | ... == ... | C.cs:120:25:120:28 | null |
| C.cs:129:13:129:38 | ... == ... | C.cs:129:14:129:29 | ... = ... |
| C.cs:129:13:129:38 | ... == ... | C.cs:129:35:129:38 | null |
| C.cs:145:13:145:39 | ... != ... | C.cs:145:14:145:30 | ... = ... |
| C.cs:145:13:145:39 | ... != ... | C.cs:145:36:145:39 | null |
| C.cs:157:16:157:24 | ... != ... | C.cs:157:16:157:16 | access to local variable s |
| C.cs:157:16:157:24 | ... != ... | C.cs:157:21:157:24 | null |
| C.cs:165:16:165:24 | ... != ... | C.cs:165:16:165:16 | access to local variable s |
| C.cs:165:16:165:24 | ... != ... | C.cs:165:21:165:24 | null |
| C.cs:172:16:172:24 | ... != ... | C.cs:172:16:172:16 | access to local variable s |
| C.cs:172:16:172:24 | ... != ... | C.cs:172:21:172:24 | null |
| C.cs:186:16:186:24 | ... != ... | C.cs:186:16:186:16 | access to local variable s |
| C.cs:211:13:211:21 | ... != ... | C.cs:211:13:211:13 | access to local variable s |
| C.cs:217:13:217:21 | ... == ... | C.cs:217:13:217:13 | access to local variable s |
| C.cs:221:13:221:21 | ... != ... | C.cs:221:13:221:13 | access to local variable s |
| C.cs:229:22:229:30 | ... != ... | C.cs:229:22:229:22 | access to local variable s |
| C.cs:235:24:235:32 | ... == ... | C.cs:235:24:235:24 | access to local variable s |
| C.cs:235:24:235:32 | ... == ... | C.cs:235:29:235:32 | null |
| D.cs:28:13:28:25 | ... != ... | D.cs:28:13:28:17 | access to parameter param |
| D.cs:38:13:38:21 | ... == ... | D.cs:38:13:38:13 | access to parameter x |
| D.cs:39:16:39:24 | ... == ... | D.cs:39:16:39:16 | access to parameter x |
| D.cs:45:13:45:22 | ... != ... | D.cs:45:13:45:14 | access to local variable o1 |
| D.cs:48:13:48:24 | ... is ... | D.cs:48:13:48:14 | access to local variable o2 |
| D.cs:51:13:51:44 | ... != ... | D.cs:51:14:51:35 | ... = ... |
| D.cs:55:23:55:32 | ... != ... | D.cs:55:23:55:24 | access to local variable o4 |
| D.cs:58:19:58:28 | ... != ... | D.cs:58:19:58:20 | access to local variable o4 |
| D.cs:59:13:59:22 | ... != ... | D.cs:59:13:59:14 | access to local variable o5 |
| D.cs:61:13:61:22 | ... != ... | D.cs:61:13:61:14 | access to local variable o4 |
| D.cs:65:14:65:29 | call to method CustomIsNull | D.cs:65:27:65:28 | access to local variable o6 |
| D.cs:69:18:69:27 | ... != ... | D.cs:69:18:69:19 | access to local variable o7 |
| D.cs:76:21:76:30 | ... == ... | D.cs:76:21:76:22 | access to local variable o8 |
| D.cs:110:26:110:35 | ... != ... | D.cs:110:26:110:27 | access to local variable xs |
| D.cs:118:13:118:21 | ... == ... | D.cs:118:13:118:13 | access to local variable x |
| D.cs:119:13:119:21 | ... == ... | D.cs:119:13:119:13 | access to local variable x |
| D.cs:127:20:127:28 | ... == ... | D.cs:127:20:127:20 | access to parameter a |
| D.cs:128:20:128:28 | ... == ... | D.cs:128:20:128:20 | access to parameter b |
| D.cs:139:13:139:21 | ... != ... | D.cs:139:13:139:13 | access to parameter a |
| D.cs:152:17:152:27 | ... != ... | D.cs:152:17:152:19 | access to parameter obj |
| D.cs:196:13:196:21 | ... == ... | D.cs:196:13:196:13 | access to local variable o |
| D.cs:206:17:206:25 | ... == ... | D.cs:206:17:206:17 | access to local variable e |
| D.cs:212:18:212:26 | ... == ... | D.cs:212:18:212:18 | access to local variable n |
| D.cs:212:18:212:26 | ... == ... | D.cs:212:23:212:26 | null |
| D.cs:216:13:216:22 | ... == ... | D.cs:216:13:216:14 | access to local variable o3 |
| D.cs:221:13:221:22 | ... == ... | D.cs:221:13:221:14 | access to local variable o4 |
| D.cs:242:13:242:25 | ... == ... | D.cs:242:13:242:17 | access to local variable other |
| D.cs:244:13:244:25 | ... != ... | D.cs:244:13:244:17 | access to local variable other |
| D.cs:336:13:336:23 | ... == ... | D.cs:336:13:336:15 | access to parameter obj |
| D.cs:341:13:341:23 | ... != ... | D.cs:341:13:341:15 | access to local variable msg |
| D.cs:367:27:367:35 | ... == ... | D.cs:367:27:367:27 | access to local variable b |
| D.cs:382:13:382:23 | ... != ... | D.cs:382:13:382:15 | access to local variable ioe |
| D.cs:390:20:390:28 | ... == ... | D.cs:390:20:390:20 | access to parameter a |
| D.cs:397:20:397:28 | ... == ... | D.cs:397:20:397:20 | access to parameter b |
| D.cs:407:14:407:22 | ... != ... | D.cs:407:14:407:14 | access to parameter x |
| D.cs:407:27:407:35 | ... == ... | D.cs:407:27:407:27 | access to parameter y |
| D.cs:407:42:407:50 | ... == ... | D.cs:407:42:407:42 | access to parameter x |
| D.cs:407:55:407:63 | ... != ... | D.cs:407:55:407:55 | access to parameter y |
| D.cs:409:13:409:21 | ... != ... | D.cs:409:13:409:13 | access to parameter x |
| D.cs:411:13:411:21 | ... != ... | D.cs:411:13:411:13 | access to parameter y |
| E.cs:10:34:10:54 | ... != ... | E.cs:10:35:10:45 | ... = ... |
| E.cs:12:32:12:52 | ... != ... | E.cs:12:33:12:43 | ... = ... |
| E.cs:20:19:20:28 | ... == ... | E.cs:20:19:20:20 | access to local variable s1 |
| E.cs:21:13:21:22 | ... == ... | E.cs:21:13:21:14 | access to local variable s2 |
| E.cs:24:19:24:28 | ... == ... | E.cs:24:19:24:20 | access to local variable s1 |
| E.cs:26:13:26:22 | ... != ... | E.cs:26:13:26:14 | access to local variable s2 |
| E.cs:70:22:70:32 | ... == ... | E.cs:70:22:70:24 | access to parameter arr |
| E.cs:83:13:83:24 | ... != ... | E.cs:83:13:83:16 | access to parameter vals |
| E.cs:85:18:85:29 | ... != ... | E.cs:85:18:85:21 | access to parameter vals |
| E.cs:153:13:153:24 | ... != ... | E.cs:153:13:153:16 | access to local variable obj2 |
| E.cs:164:17:164:25 | ... == ... | E.cs:164:17:164:17 | access to parameter a |
| E.cs:175:19:175:29 | ... == ... | E.cs:175:19:175:21 | access to parameter obj |
| E.cs:176:13:176:22 | ... == ... | E.cs:176:13:176:14 | (...) ... |
| E.cs:180:13:180:23 | ... == ... | E.cs:180:13:180:15 | access to parameter obj |
| E.cs:184:13:184:22 | ... == ... | E.cs:184:13:184:14 | (...) ... |
| E.cs:193:17:193:17 | access to parameter o | E.cs:193:17:193:17 | access to parameter o |
| E.cs:208:13:208:23 | ... is ... | E.cs:208:13:208:13 | access to parameter s |
| E.cs:245:13:245:22 | access to property HasValue | E.cs:245:13:245:13 | access to parameter i |
| E.cs:252:13:252:21 | ... != ... | E.cs:252:13:252:13 | access to parameter i |
| E.cs:259:13:259:21 | ... == ... | E.cs:259:13:259:13 | access to parameter i |
| E.cs:274:17:274:25 | ... != ... | E.cs:274:17:274:17 | access to local variable o |
| E.cs:284:9:284:9 | access to local variable o | E.cs:284:9:284:9 | access to local variable o |
| E.cs:293:13:293:13 | access to local variable s | E.cs:293:13:293:13 | access to local variable s |
| E.cs:306:31:306:31 | access to field l | E.cs:306:31:306:31 | access to field l |
| E.cs:309:13:309:22 | access to property HasValue | E.cs:309:13:309:13 | access to field l |
| E.cs:321:13:321:30 | ... is ... | E.cs:321:14:321:21 | ... ?? ... |
| E.cs:321:14:321:15 | access to parameter s1 | E.cs:321:14:321:15 | access to parameter s1 |
| E.cs:330:17:330:28 | (...) ... | E.cs:330:17:330:28 | (...) ... |
| E.cs:336:17:336:17 | access to parameter s | E.cs:336:17:336:17 | access to parameter s |
| E.cs:342:17:342:17 | access to parameter s | E.cs:342:17:342:17 | access to parameter s |
| E.cs:348:21:348:21 | access to parameter s | E.cs:348:21:348:21 | access to parameter s |
| E.cs:354:21:354:21 | access to parameter s | E.cs:354:21:354:21 | access to parameter s |
| E.cs:355:13:355:21 | dynamic call to operator != | E.cs:355:13:355:13 | access to local variable x |
| E.cs:361:17:361:17 | access to parameter s | E.cs:361:17:361:17 | access to parameter s |
| E.cs:362:13:362:29 | ... != ... | E.cs:362:13:362:13 | access to local variable x |
| E.cs:382:14:382:23 | ... == ... | E.cs:382:14:382:15 | access to parameter e1 |
| E.cs:382:28:382:37 | ... != ... | E.cs:382:28:382:29 | access to parameter e2 |
| E.cs:382:44:382:53 | ... != ... | E.cs:382:44:382:45 | access to parameter e1 |
| E.cs:382:58:382:67 | ... == ... | E.cs:382:58:382:59 | access to parameter e2 |
| E.cs:384:13:384:22 | ... == ... | E.cs:384:13:384:14 | access to parameter e1 |
| E.cs:384:27:384:36 | ... == ... | E.cs:384:27:384:28 | access to parameter e2 |
| E.cs:391:9:391:9 | access to parameter i | E.cs:391:9:391:9 | access to parameter i |
| E.cs:397:9:397:13 | access to parameter color | E.cs:397:9:397:13 | access to parameter color |
| E.cs:404:9:404:9 | access to local variable i | E.cs:404:9:404:9 | access to local variable i |
| E.cs:411:9:411:9 | access to local variable i | E.cs:411:9:411:9 | access to local variable i |
| E.cs:422:13:422:22 | access to property HasValue | E.cs:422:13:422:13 | access to parameter i |
| E.cs:429:13:429:22 | access to property HasValue | E.cs:429:13:429:13 | access to parameter i |
| E.cs:437:23:437:31 | ... is ... | E.cs:437:23:437:23 | access to parameter s |
| E.cs:447:23:447:35 | ... is ... | E.cs:447:23:447:23 | access to parameter s |
| Forwarding.cs:45:66:45:75 | call to method IsNull | Forwarding.cs:45:66:45:66 | access to parameter o |
| Forwarding.cs:59:13:59:21 | ... == ... | Forwarding.cs:59:13:59:13 | access to parameter o |
| Forwarding.cs:78:16:78:39 | call to method ReferenceEquals | Forwarding.cs:78:32:78:32 | access to parameter o |
| Forwarding.cs:83:16:83:24 | ... != ... | Forwarding.cs:83:16:83:16 | access to parameter o |
| GuardedString.cs:19:13:19:13 | access to local variable s | GuardedString.cs:19:13:19:13 | access to local variable s |
| GuardedString.cs:22:13:22:13 | access to local variable s | GuardedString.cs:22:13:22:13 | access to local variable s |
| GuardedString.cs:25:13:25:13 | access to local variable s | GuardedString.cs:25:13:25:13 | access to local variable s |
| GuardedString.cs:28:13:28:13 | access to local variable s | GuardedString.cs:28:13:28:13 | access to local variable s |
| GuardedString.cs:31:13:31:13 | access to local variable s | GuardedString.cs:31:13:31:13 | access to local variable s |
| GuardedString.cs:34:13:34:13 | access to local variable s | GuardedString.cs:34:13:34:13 | access to local variable s |
| NullAlwaysBad.cs:9:17:9:25 | ... != ... | NullAlwaysBad.cs:9:17:9:17 | access to parameter s |
| NullAlwaysGood.cs:9:17:9:25 | ... != ... | NullAlwaysGood.cs:9:17:9:17 | access to parameter s |
| NullMaybeGood.cs:7:13:7:21 | ... != ... | NullMaybeGood.cs:7:13:7:13 | access to parameter o |
